tensorflow/core/grappler/optimizers/remapper.cc

```diff
@@ -968,6 +980,111 @@ bool FindFusedBatchNormEx(const RemapperContext& ctx, int node_index,
   return false;
 }
 
+bool FindFusedBatchNormGradEx(const RemapperContext& ctx, int node_index,
+                              FusedBatchNormGradEx* matched) {
+  // Root of the pattern must be a FusedBatchNormGrad.
+  const auto* node_view = ctx.graph_view.GetNode(node_index);
+
+	// Returns true iff the node is a compatible FusedBatchNormGrad node.
+  const auto valid_batch_norm_grad =
+      [&](const utils::MutableNodeView& fused_batch_norm_grad) -> bool {
+    const auto* node_def = fused_batch_norm_grad.node();
+    if (!IsFusedBatchNormGrad(*node_def) ||
+        HasControlFaninOrFanout(fused_batch_norm_grad)) return false;
+
+    // We fuse FusedBatchNormGrad on GPU.
+    if (!NodeIsOnGpu(node_def)) return false;
+
+    // Data type must be DT_HALF.
+    DataType t_dtype = GetDataTypeFromAttr(*node_def, "T");
+    if (t_dtype != DT_HALF) return false;
+
+    // We rely on cuDNN for computing FusedBatchNormGrad with side
+    // outputs and activation. cuDNN only supports NHWC data layout.
+    string data_format;
+    if (!GetNodeAttr(*node_def, kDataFormat, &data_format).ok()) return false;
+    if (data_format != "NHWC") return false;
+
+    // Channel dimension must be a multiple of 4.
+    const auto& props = ctx.graph_properties.GetInputProperties(
+        node_def->name());
+    const bool valid_channel_dim = !props.empty() &&
+                                   props[0].shape().dim_size() == 4 &&
+                                   props[0].shape().dim(3).size() % 4 == 0;
+    if (!valid_channel_dim) return false;
+
+    // cuDNN must support CUDNN_BATCHNORM_SPATIAL_PERSISTENT mode.
+    if (!BatchnormSpatialPersistentEnabled()) return false;
+
+    // FusedBatchNormV2 and V3 have an extra type parameter.
+    if (node_def->op() != "FusedBatchNorm" &&
+        !HasDataType(node_def, DT_FLOAT, "U")) return false;
+
+    return true;
+  };
+
+  if (!valid_batch_norm_grad(*node_view)) return false;
+
+  if (node_view->NumRegularFanins() < 1) return false;
+
+  const auto& regular_fanin_0 = node_view->GetRegularFanin(0);
+  const auto* relugrad_node_view = regular_fanin_0.node_view();
+  const auto* relugrad_node_def = relugrad_node_view->node();
+  bool is_relugrad = IsReluGrad(*relugrad_node_def);
+
+  if (!is_relugrad || HasControlFaninOrFanout(*relugrad_node_view) ||
+      IsInPreserveSet(ctx, relugrad_node_def)) return false;
+
+  if (relugrad_node_view->NumRegularFanins() < 1) return false;
+  // Find its corresponding forward node. We need the node to determine if the
+  // type is bn+add+act or bn+act. Also, we need to access its "offset" input.
+  const auto& fanin_1 = relugrad_node_view->GetRegularFanin(1);
+  const auto* fwd_node_view = fanin_1.node_view();
+  const auto* fwd_node_def = fwd_node_view->node();
```

Comment:
This seems to be an unused variable. This blocks merging, because internally we treat unused variables as an error.

Reply:
I'll try to fix it internally, so no need to update the PR right now.